ShareX.ScreenCaptureLib/Shapes/Tool/CutOutTool.cs

```diff
@@ -0,0 +1,203 @@
+                    {
+                        buttonsLeft = Manager.Form.ClientArea.Left + buttonOffset;
+                    }
+                    else if (overflowsRight)
+                    {
+                        buttonsRight = Manager.Form.ClientArea.Right - buttonOffset;
+                        buttonsLeft = buttonsRight - 2 * buttonSize.Width - buttonOffset;
+                    }
+                    confirmButton.Rectangle = new RectangleF(buttonsLeft, buttonsTop, buttonSize.Width, buttonSize.Height);
+                    cancelButton.Rectangle = confirmButton.Rectangle.LocationOffset(buttonSize.Width + buttonOffset, 0);
+                }
+                else
+                {
+                    float spaceRight = Manager.Form.ClientArea.Right - Rectangle.Right;
+                    bool positionRight = spaceRight >= buttonSize.Width + 2 * buttonOffset;
+                    float buttonsLeft = positionRight ? Rectangle.Right + buttonOffset : Rectangle.Left - buttonOffset - buttonSize.Width;
+                    float buttonsTop = Rectangle.Top + Rectangle.Height / 2 - (2 * buttonSize.Height + buttonOffset) / 2;
+                    float buttonsBottom = buttonsTop + 2 * buttonSize.Height + buttonOffset;
+                    bool overflowsTop = buttonsTop < Manager.Form.ClientArea.Top + buttonOffset;
+                    bool overflowsBottom = buttonsBottom >= Manager.Form.ClientArea.Bottom - buttonOffset;
+                    if (overflowsTop && overflowsBottom)
+                    {
+                        // can't fix
+                    }
+                    else if (overflowsTop)
+                    {
+                        buttonsTop = Manager.Form.ClientArea.Top + buttonOffset;
+                    }
+                    else if (overflowsBottom)
+                    {
+                        buttonsBottom = Manager.Form.ClientArea.Bottom - buttonOffset;
+                        buttonsTop = buttonsBottom - 2 * buttonSize.Height - buttonOffset;
+                    }
+                    confirmButton.Rectangle = new RectangleF(buttonsLeft, buttonsTop, buttonSize.Width, buttonSize.Height);
+                    cancelButton.Rectangle = confirmButton.Rectangle.LocationOffset(0, buttonSize.Height + buttonOffset);
+                }
+            }
+        }
+
+        public override void OnDraw(Graphics g)
+        {
+            using (Image selectionHighlightPattern = ImageHelpers.CreateCheckerPattern(8, 8, Color.FromArgb(128, Color.White), Color.FromArgb(128, Color.Gray)))
+            using (Brush selectionHighlightBrush = new TextureBrush(selectionHighlightPattern, System.Drawing.Drawing2D.WrapMode.Tile))
+            {
+                if (IsHorizontalTrim)
+                {
+                    g.FillRectangle(selectionHighlightBrush, new RectangleF(Rectangle.X, Math.Max(g.ClipBounds.Y, Manager.Form.CanvasRectangle.Y), Rectangle.Width, Math.Min(g.ClipBounds.Height, Manager.Form.CanvasRectangle.Height)));
```

Comment:
These rectangles can be stored as property such as:

```cs
        public Rectangle CutOutRectangle
        {
            get
            {
                ...
            }
        }
```

That way you can also use it inside `OnUpdate()` method while positioning buttons properly.

Reply:
That does make the code a bit neater, but I don't see how it'd help in OnUpdate(). The button positioning barely concerns itself with the specific shape of the entire cut out area, just the two extents in the direction being cut, and those are the same for the user selection and the displayed cut. I think using the user selection (`Rectangle` property) is better for the buttons, since that will typically position them closer to where the user releases the mouse.

The OnUpdate() button positioning could be shortened a bit, but that would make it harder to read/follow, so I think I prefer keeping it as-is.